gpytorch/utils/trace.py

```diff
@@ -0,0 +1,134 @@
+    since `trace( A^T B ) = sum(A * B)`
+
+    However, if `A` and `B` are more efficiently represented by matrix multiplication closures
+    (as is the case if A is a  Toeplitz matrices, if A is the inverse of a PSD matrix, etc.)
+    then this will be far more efficient, since `A' and `B'` will have significantly lower dimensionality.
+
+    Args:
+        - left_matmul_closure (Tensor or Variable or function(Tensor) -> Tensor or
+            function(Variable) -> Variable) defines the left matrix, `A` (n x m), or the
+            matrix multiplcation function f(X) = AX
+
+            left_matmul_closure can also be None, which specifies that A is the identity matrix
+
+        - right_matmul_closure (Tensor or Variable or function(Tensor) -> Tensor or
+            function(Variable) -> Variable) defines the right matrix, `B` (n x m), or the
+            matrix multiplcation function f(X) = BX
+
+            right_matmul_closure can also be None, which specifies that B is the identity matrix
+
+        - size (int) -> the number of columns in the matrices `A` and `B` (i.e. m).
+            Only required if left_matmul_closure and right_matmul_closures are not Tensors/Variables
+
+        - num_samples (int) -> the number of random variable samples to stochastically estimate
+            trace. Defaults to `gpytorch.functions.num_trace_samples`
+
+        - estimator_type (str) -> Options are 'mub' (Mutually unbiased bases) or
+            'hutchinson' (Rademacher random variables). (Default 'mub')
+
+        - use_vars (bool) -> If False, then operations are performed on Tensors.
+            If True, then performs operations on Variables. (Default False)
+    """
+    if torch.is_tensor(left_matmul_closure):
+        left_matrix = left_matmul_closure
+        size = left_matrix.size(1)
+        tensor_cls = type(left_matrix)
+        left_matmul_closure = left_matrix.matmul
+
+    if torch.is_tensor(right_matmul_closure):
+        right_matrix = right_matmul_closure
+        size = right_matrix.size(1)
+        tensor_cls = type(right_matrix)
+        right_matmul_closure = right_matrix.matmul
+
+    if left_matmul_closure is None:
+        left_matmul_closure = _identity
+
+    if right_matmul_closure is None:
+        right_matmul_closure = _identity
+
+    if size is None:
+        raise RuntimeError('Size must be specified, since neither left_matmul_closure nor'
+                           ' right_matmul_closure are Tensors/Variables')
+
+    # Default num_samples, tensor_cls
+    if num_samples is None:
+        num_samples = gpytorch.functions.num_trace_samples
+
+    if tensor_cls is None:
+        tensor_cls = torch.Tensor
+
+    # Return A and B if we're using deterministic mode
+    if not gpytorch.functions.fastest or size < num_samples:
+        eye = tensor_cls(size).fill_(1).diag()
+        if use_vars:
+            eye = Variable(eye)
+        return left_matmul_closure(eye), right_matmul_closure(eye)
```

Comment:
This line right here is the problem (or an example of it). If we need to compute Tr(K^{-1}), left_matmul_closure(eye) will return an n-by-n matrix, which destroys the linear space complexity of KISS-GP. This can be solved by doing it in large pieces instead of all at once, e.g. estimating the trace with left_matmul_closure(eye[:, :20]), left_matmul_closure(eye[:, 20:40]), etc by performing the final multiply-and-sum parts one at a time.

Reply:
Or in general with the stochastic version the space complexity is O(ns). In principle we can just average over 10-20 probe vectors, then average over the next 10-20, and so on, rather than computing with all s at once.

Ruihan and I simply think we should be able to set a batch size so that when n is very large the space requirement doesn't get out of control.

Reply:
I still don't understand. We'll only enter this loop if the matrix is smaller than 20x20 (or if the user specifies not to do stochastic trace estimation).

Reply:
And what is the purpose of batching? I think trying to implement a batched version will probably be significantly slower than what's here right now.

Reply:
Our purpose is to make space complexity in our control. The space complexity is O(sn) now and it will be O(n^2) in worst case, where n is the number of training data in exact likelihood in regression task.